### PR TITLE
LibGL: Implement texture sampling and OpenGL texture state management

### PR DESCRIPTION
**Pull request by Quaker762**
Going to keep this open as a WIP branch for now while I research the best way to do this. Feedback most certainly welcome!

The goal here is to add texture allocation and state management to `LibGL` then add texturing sampling to the renderer.

### CODE REVIEW

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 1]
```diff
@@ -628,6 +628,37 @@ void SoftwareGLContext::gl_disable(GLenum capability)
...
+        // We're out of free previously allocated names. Let's allocate a new contiguous amount from the
+        // last known texture name
+        // 'n' now is the remaining number of names to allocate
+        for (auto name = m_last_texture_id; name < m_last_texture_id + n; name++) {
+            textures[tex_array_index++] = name;
+        }
+
+        m_last_texture_id += n;
+    }
+}
```
Comment (ccapitalK) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:652]: This code is a bit hard to read, I think it would be better to do it as:

```suggestion
void SoftwareGLContext::gl_gen_textures(GLsizei n, GLuint *textures) {
    for (size_t i = 0; i < n; ++i) {
        if (!m_free_textures_names.is_empty()) {
            // We have an already allocated free texture name. Pop it into the array
            textures[i] = m_free_texture_names.dequeue();
        } else {
            // We're out of allocated texture names, generate a new one
            textures[i] = m_last_texture_id++;
        }
    }
}
```

Comment (predmond) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:652]: Out of curiosity, is it even necessary to reuse texture names?

Comment (sunverwerth) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:652]: https://www.khronos.org/registry/OpenGL/specs/gl/glspec15.pdf

The 1.5 spec for `glGenTextures` states `returns n previously  unused  texture  object  names  in textures.   These  names  are marked as used, for the purposes of GenTextures only, but they acquire texture state  and  a  dimensionality  only  when  they  are  first  bound,  just  as  if  they  were unused.`

And for `glDeleteTextures` `After a texture object is deleted, it has no contents or dimensionality, and its name is again unused.  `

So I guess it's not strictly necessary to reuse texture names?

Comment (ccapitalK) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:652]: TBH, I think it would be sufficient to just have an atomic counter, the only issue is if you have a very long running program that constantly asks for texture names.

Comment (sunverwerth) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:652]: FWIW, I just tested creating and deleting 10 textures in a loop and always got the same 10 texture names back. Windows 10, NVIDIA GTX 1080.

Comment (Quaker762) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:691]: > Out of curiosity, is it even necessary to reuse texture names?

Tehcnically, no, but as @sunverwerth and @ccapitalK have said, it would be breaking the spec + creating behaviour that isn't seen on other OpenGL implementations.

File: AK/Stack.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        return m_stack.size();
+    }
+
+    bool pop()
+    {
+        if (is_empty())
+            return false;
+
+        m_stack.resize(m_stack.size() - 1, true);
```
Comment (majestrate) [inline AK/Stack.h:54]: `m_stack.resize_and_keep_capacity(m_stack.size() - 1)` maybe?

File: Userland/Libraries/LibGL/Tex/Texture.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+{
+public:
+    Texture() = default;
+    ~Texture();
+
+    void upload_texture_data(GLint format, GLint internal_format, const GLvoid* pixel_data);
+
+private:
+    template<typename TCallback>
+    void swizzle(const Vector<u8>& pixels, TCallback&& callback);
```
Comment (alimpfard) [inline Userland/Libraries/LibGL/Tex/Texture.h:25]: This won't work for any kind of `TCallback`, the definition in the cpp file is not visible to other translation units.
you can either specialise them here by forward-declaring them (`template void swizzle<SomeTypeForTCallback>(...);`), or just bring the definition into this header x)

File: Userland/Libraries/LibGL/Tex/NameAllocator.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+void TextureNameAllocator::allocate(GLsizei count, GLuint *textures)
+{
+    for (auto i = 0; i < count; ++i) {
+        if (!m_free_texture_names.is_empty()) {
+            textures[i] = m_free_texture_names.top();
+            m_free_texture_names.pop();
+        } else {
+            // We're out of free previously allocated names. Let's allocate a new contiguous amount from the
+            // last known texture name
+            textures[i] = m_last_texture_id++;
```
Comment (sunverwerth) [inline Userland/Libraries/LibGL/Tex/NameAllocator.cpp:21]:
```suggestion
            textures[i] = ++m_last_texture_id;
```
The first valid name is 1, not 0. So either preincrement here or initialize m_last_texture_id with 0.

**Review by sunverwerth — CHANGES_REQUESTED** [review]

File: AK/Stack.h (hunk 2 above)
Comment (Quaker762) [reply AK/Stack.h:54]: Good spot!

File: Userland/Libraries/LibGL/Tex/Texture.h (hunk 3 above)
Comment (Quaker762) [reply Userland/Libraries/LibGL/Tex/Texture.h:25]: Going to remove the implementation of this for the moment until I flesh out the rest of the texture class data members. We can iterate on this design later in the week on Discord :P

File: Userland/Libraries/LibGL/Tex/NameAllocator.cpp (hunk 4 above)
Comment (Quaker762) [reply Userland/Libraries/LibGL/Tex/NameAllocator.cpp:21]: Good spot! I've changed the initialised value of `m_last_texture_id` to `1`

File: AK/Stack.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    bool pop()
+    {
+        if (is_empty())
+            return false;
+
+        m_stack.resize_and_keep_capacity(m_stack.size() - 1);
+        return true;
+    }
+
+    T& top()
```
Comment (majestrate) [inline AK/Stack.h:58]: would it make sense to have a `const` version of `top()` as well?

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 6]
```diff
@@ -732,6 +732,36 @@ void SoftwareGLContext::gl_disable(GLenum capability)
         m_rasterizer.set_options(rasterizer_options);
 }
 
+void SoftwareGLContext::gl_gen_textures(GLsizei n, GLuint* textures)
+{
+    if (n < 0) {
```
Comment (majestrate) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:737]: should this be `n <= 0` instead? is a zero value here valid?

Comment (majestrate) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:737]: also what about a null check for `textures` ?

File: Userland/Libraries/LibGL/Tex/NameAllocator.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        }
+    }
+}
+
+void TextureNameAllocator::free(GLsizei count, const GLuint *textures)
```
Comment (majestrate) [inline Userland/Libraries/LibGL/Tex/NameAllocator.cpp:26]: `nullptr` check could be useful here for `textures`

File: Userland/Libraries/LibGL/Tex/Texture.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    // Make sure user passed in support pixel data type
+    if (type != GL_UNSIGNED_BYTE) {
+        dbgln("LibGL: Unsupported pixel format type {}", type);
+        VERIFY_NOT_REACHED();
+    }
+
+    // Somebody passed us in nullptr...
+    if (pixels == nullptr) {
+        dbgln("LibGL: pixels == nullptr when uploading texture data.");
+        return;
```
Comment (majestrate) [inline Userland/Libraries/LibGL/Tex/Texture.cpp:35]: this seems like a great place to throw an `std::invalid_argument` exception but it seems that serenity os doesn't like using exceptions `:<`

Comment (alimpfard) [reply Userland/Libraries/LibGL/Tex/Texture.cpp:35]: Exceptions are disabled, and will most likely remain disabled : )

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp (hunk 6 above)
Comment (Quaker762) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:737]: Nope. If `n == 0`, nothing happens. This will just fall through. 

image

File: Userland/Libraries/LibGL/Tex/NameAllocator.cpp (hunk 7 above)
Comment (Quaker762) [reply Userland/Libraries/LibGL/Tex/NameAllocator.cpp:26]: OpenGL does no checks on pointers. If somebody passes in a bad value and hits `SIGSEGV`, that's their problem :P

File: AK/Stack.h (hunk 5 above)
Comment (Quaker762) [reply AK/Stack.h:58]: Good point, I'll add that in now :P

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp (hunk 6 above)
Comment (majestrate) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:737]: so when n is zero does `m_name_allocator.allocate(n, textures);` work as expected?

Comment (Quaker762) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:737]: Yes.

```
for (auto i = 0; i < 0; ++i)
```

**Comment by Quaker762** [comment]
The CI seems to be lying here... Line 1250 of `SoftwareGLContext.cpp` is currently type `int`, not type `size_t`....

https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGL/SoftwareGLContext.cpp#L1250

File: AK/Stack.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        return m_stack.last();
+    }
+
+    const T& top() const
+    {
+        return m_stack.last();
+    }
+
+private:
+    Vector<T, stack_size> m_stack;
```
Comment (alimpfard) [inline AK/Stack.h:67]: Maybe I'm being silly, but can we instead switch this to fixed storage?
Vector has some mystifying behaviours and errors:
> what if you wanted to write code, but the compiler said
```c++
../.././AK/Vector.h:60:34: error: '*(AK::Vector<Wasm::Value, 0>*)((char*)&<unnamed> + offsetof(AK::EntryType, AK::Variant<Wasm::Value, Wasm::Label, Wasm::Frame>::m_data[4])).AK::Vector<Wasm::Value>::m_outline_buffer' may be used uninitialized [-Werror=maybe-uninitialized]
   60 |         , m_outline_buffer(other.m_outline_buffer)
      |                            ~~~~~~^~~~~~~~~~~~~~~~
```

> and then you checked and saw
```
T* m_outline_buffer { nullptr };
```

File: Userland/Libraries/LibGL/SoftwareGLContext.h [hunk 10]
```diff
@@ -6,6 +6,7 @@
 
 #pragma once
 
+#include "Tex/NameAllocator.h"
```
Comment (alimpfard) [inline Userland/Libraries/LibGL/SoftwareGLContext.h]: This doesn't seem to be related to this commit?

File: Userland/Libraries/LibGL/GLTexture.cpp [hunk 11]
```diff
@@ -4,8 +4,8 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
-#include <LibGL/GL/gl.h>
 #include "GLContext.h"
+#include <LibGL/GL/gl.h>
```
Comment (alimpfard) [inline Userland/Libraries/LibGL/GLTexture.cpp:8]: Clang-format did this? W e i r d 

(The fixed position is correct, what it was at...not so much)

**Review by alimpfard — COMMENTED** [review]
Just some nits, otherwise LGTM

File: Userland/Libraries/LibGL/GLTexture.cpp (hunk 11 above)
Comment (Quaker762) [reply Userland/Libraries/LibGL/GLTexture.cpp:8]: I might've run `clang-format` on this in the wrong commit... Whoops... I'll have to comb through and get that fixed..

File: AK/Stack.h (hunk 9 above)
Comment (Quaker762) [reply AK/Stack.h:67]: As per Discord conversation, this is a refactor for another time :)
